### ISSUE TITLE
driver picks lgcc

### ISSUE THREAD

**Issue by androm3da** [ClangBuiltLinux/linux#759]
Hexagon clang driver uses `-lgcc`, we should be able to use `clang_rt.builtins` instead.

```
ld.lld -r -o vmlinux.o --whole-archive arch/hexagon/kernel/head.o init/built-in.a usr/built-in.a arch/hexagon/kernel/built-in.a arch/hexagon/mm/built-in.a arch/hexagon/lib/built-in.a kernel/built-in.a certs/built-in.a mm/built-in.a fs/built-in.a ipc/built-in.a security/built-in.a crypto/built-in.a block/built-in.a lib/built-in.a drivers/built-in.a sound/built-in.a net/built-in.a virt/built-in.a --no-whole-archive --start-group lib/lib.a libgcc.a --end-group
ld.lld: error: cannot open libgcc.a: No such file or directory

```

**Comment by nathanchance** [comment]
Should be easy enough to solve with `-rtlib=compiler-rt` when using Clang.

How are the builtins created, since they are not created as part of the normal build process it seems?

**Comment by androm3da** [comment]
Sure we can override it but I want the driver to DTRT, I figure we will need this for the buildroot to work without the override too.  We have a guard for `Triple::Linux` in the downstream Hexagon driver that we are missing upstream.

We might have downstream changes to `compiler-rt` that need to get pushed upstream too.

**Comment by nickdesaulniers** [comment]
I didn't think that libgcc_s nor compiler-rt gets used for any other arch's kernel builds.  Do you know what functions from libgcc_s are relied upon from the hexagon port?

**Comment by nathanchance** [comment]
By commenting out the libgcc link, looks like:

```
ld.lld: error: undefined symbol: __hexagon_memcpy_likely_aligned_min32bytes_mult8bytes
ld.lld: error: undefined symbol: __hexagon_udivsi3
ld.lld: error: undefined symbol: __hexagon_umodsi3
ld.lld: error: undefined symbol: __hexagon_divsi3
ld.lld: error: undefined symbol: __hexagon_modsi3
```

**Comment by androm3da** [comment]
Yes: optimized memcpy, integer modulus and division.

**Comment by nickdesaulniers** [comment]
Other arch's end up copying implementations from libgcc wholesale into the kernel. This leads to divergence over time, as bugs get fixed in libgcc and the copy in the kernel is now outdated.  (But it does simplify the build in that you don't need a libgcc for your target arch installed to cross compile the kernel).

Does compiler_rt implement all of these routines for hexagon?

Would it be easier to copy these routines into the kernel sources?

**Comment by androm3da** [comment]
> Would it be easier to copy these routines into the kernel sources?

I'm willing to consider that.  But we have had a working configuration downstream like this for some time, so I'd like to try to reproduce that upstream first.

> But it does simplify the build in that you don't need a libgcc for your target arch installed to cross compile the kernel

Yes, it's an additional step to build compiler-rt but I'd be willing to add the patch to ClangBuiltLinux to make it work, if it's acceptable.

**Comment by nickdesaulniers** [comment]
I'm not against carrying out of tree patches, as long the number of such patches never grows beyond 10 and we have a plan to upstream them. Though for this bug, I'm not sure yet that's necessary.

In regards to the bug, maybe the compiler lib should be chosen based on `CONFIG_CC_IS_CLANG` vs `CONFIG_CC_IS_GCC`.

**Comment by nickdesaulniers** [comment]
I spoke with some of the RISC-V stuff about their libgcc routines, and they as well suggested copying the source implementations from libgcc into the kernel; this is common practice for most other arch's AFAIK.

**Comment by androm3da** [comment]
> I spoke with some of the RISC-V stuff about their libgcc routines, and they as well suggested copying the source implementations from libgcc into the kernel; this is common practice for most other arch's AFAIK.

Okay, this makes sense.  I will work on it.

**Comment by nickdesaulniers** [comment]
sent a post to our [mailing list](https://groups.google.com/g/clang-built-linux/c/OvG71CL1KWc), linux-arch@vger.kernel.org, and linux-hexagon@vger.kernel.org asking if anyone was planning to support this.

**Comment by androm3da** [comment]
Candidate fixes in git://git.kernel.org/pub/scm/linux/kernel/git/bcain/linux using `for-linus` branch, commit 43810cfdd2e6af9c57c0975e1b920f2b18c484b6.

Built with an llvm+clang 11.0.0 cross toolchain.  I haven't tried 12.0.0 or trunk yet, but will soon.

(BTW my build used `LLVM_IAS=1` - because `no-integrated-as` encountered a toolchain bug.  I reproduced it on trunk, so I'll open that bug at https://bugs.llvm.org/ soon).

**Comment by nickdesaulniers** [comment]
Fix: https://git.kernel.org/pub/scm/linux/kernel/git/bcain/linux.git/commit/?h=for-linus&id=43810cfdd2e6af9c57c0975e1b920f2b18c484b6

**Comment by nickdesaulniers** [comment]
f1f99adf05f2138ff2646d756d4674e302e8d02d and friends landed in v5.13-rc1. thanks all!
